### PR TITLE
Improve hardcoded domains and IP addresses

### LINKED ISSUE #1549: Drop dependence on lists of external IP addresses/domain names
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

There are currently a few places where we maintain lists of external IP addresses (eg. the IP addresses of Ubuntu's package servers) or domain names (eg. lists of AzureAD authentication endpoints).

We should try, wherever possible, to replace these with generic wildcarded "allow-all" rules as long as this doesn't cause security issues.

#### Firewall rules
- `AllowExternalAzureAutomationOperations` ApplicationRule could be replaced with a NetworkRule that allows the `GuestAndHybridManagement` service tag (NB. this will need testing, as we haven't verified what this tag allows)
- `AllowExternalAzureADConnectPasswordReset` ApplicationRule could potentially allow `*..servicebus.windows.net` (which is a security risk, but this rule only applies to the DC)
- `AllowExternalAzureADConnectOperations` ApplicationRule could potentially allow `*.blob.core.windows.net` (which is a *large* risk, but this rule only applies to the DC)

#### NSG rules
- `AllowAzureAutomationOutbound` rule on the compute NSG could possibly use the `GuestAndHybridManagement` service tag instead of hard-coded IP addresses

## :steam_locomotive: How could this be done?

Going through the code carefully and evaluating each potential change

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).
- [x] You have formatted your code using appropriate automated tools (for example `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>` for Powershell).

### :closed_umbrella: Related issues

Closes #1549

### :microscope: Tests

Applied rules to an SHM/SRE _in situ_
Unable to find an obvious holes _e.g._,
  - Accessing domains on the internet
  - Accessing public or private storage accounts with a permissive SAS token (`*.blob.core.windows.net`)

### CODE REVIEW

File: deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json [hunk 1]
```diff
@@ -24,6 +24,19 @@
...
                 "rules": [
+                    {
+                        "name": "AllowExternalAzureAutomationOperations",
+                        "protocols": [
+                            "HTTPS:443"
+                        ],
+                        "targetAddresses": [
+                            "GuestAndHybridManagement",
+                            "AzureMonitor"
+                        ],
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json]: Do we know what is covered by these two tags? Do either of them include any domains that users can upload to (e.g. Blob Storage, Log Analytics, etc.)?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-external-tier3.json [hunk 2]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-external-tier3.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-internal-tier2.json [hunk 3]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-internal-tier2.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-internal-tier3.json [hunk 4]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-internal-tier3.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-proxies-tier2.json [hunk 5]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-proxies-tier2.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-proxies-tier3.json [hunk 6]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-proxies-tier3.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-update-servers.json [hunk 7]
```diff
@@ -95,11 +95,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-update-servers.json:98]: Do you also need a rule for `AzureMonitor` here?

File: deployment/secure_research_environment/network_rules/sre-nsg-rules-compute.json [hunk 8]
```diff
@@ -197,11 +197,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/secure_research_environment/network_rules/sre-nsg-rules-compute.json:200]: Do you also need a rule for `AzureMonitor` here?

File: deployment/secure_research_environment/network_rules/sre-nsg-rules-databases.json [hunk 9]
```diff
@@ -149,11 +149,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/secure_research_environment/network_rules/sre-nsg-rules-databases.json:152]: Do you also need a rule for `AzureMonitor` here?

File: deployment/secure_research_environment/network_rules/sre-nsg-rules-guacamole.json [hunk 10]
```diff
@@ -143,11 +143,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/secure_research_environment/network_rules/sre-nsg-rules-guacamole.json:146]: Do you also need a rule for `AzureMonitor` here?

File: deployment/secure_research_environment/network_rules/sre-nsg-rules-webapps.json [hunk 11]
```diff
@@ -115,11 +115,7 @@
         "name": "AllowAzureAutomationOutbound",
         "access": "Allow",
         "description": "Allow outbound connections to Azure automation servers",
-        "destinationAddressPrefix": [
-            {{#shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-                "{{.}}",
-            {{/shm.monitoring.updateServers.externalIpAddresses.azureAutomation}}
-        ],
+        "destinationAddressPrefix": "GuestAndHybridManagement",
```
Comment (jemrobinson) [inline deployment/secure_research_environment/network_rules/sre-nsg-rules-webapps.json:118]: Do you also need a rule for `AzureMonitor` here?

File: deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json (hunk 1 above)
Comment (JimMadge) [reply deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json]: Not beyond [this](https://learn.microsoft.com/en-us/azure/virtual-network/service-tags-overview#available-service-tags).

> GuestAndHybridManagement 	Azure Automation and Guest Configuration.
> AzureMonitor 	Log Analytics, Application Insights, AzMon, and custom metrics (GiG endpoints).
> 
> Note: For Log Analytics, the Storage tag is also required. If Linux agents are used, GuestAndHybridManagement tag is also required.

Also found [this lookup for service tags](https://azservicetags.azurewebsites.net/servicetag), not sure if that is helpful.

File: deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-external-tier3.json (hunk 2 above)
Comment (JimMadge) [reply deployment/safe_haven_management_environment/network_rules/shm-nsg-rules-repository-mirrors-external-tier3.json:118]: I don't think so. Actually, I don't think it is needed above either.

File: deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json (hunk 1 above)
Comment (jemrobinson) [reply deployment/safe_haven_management_environment/network_rules/shm-firewall-rules.json]: Can we do some tests? Deploy with these settings, deploy a couple of resources that users can upload to (e.g. storage accounts, log analytics etc.) and check whether they can be accessed?

**Comment by JimMadge** [comment]
@jemrobinson @craddm I've applied these rules to an SHM/SRE. I didn't find any obvious problems, so I would like to propose merging this and plan to focus on this in the upcoming pen test.

Could you add your reviews?

**Review by jemrobinson — APPROVED** [review]
I suggest merging this but making a new issue under the 4.2.0 milestone along the lines of "confirm that this works following pentesting".

**Comment by JimMadge** [comment]
I might not open an issue _yet_ but I'm planning to ask to test this in particular during the pen test.

Maybe we can continue to use this issue to describe what we have done in the open?

**Pull request merged**